File: src/org/nextras/orm/intellij/inspection/ReadOnlyPropertyInspection.java
package org.nextras.orm.intellij.inspection;

import com.intellij.codeInspection.LocalQuickFix;
import com.intellij.codeInspection.ProblemDescriptor;
import com.intellij.codeInspection.ProblemsHolder;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElementVisitor;
import com.jetbrains.php.PhpIndex;
import com.jetbrains.php.lang.documentation.phpdoc.psi.PhpDocProperty;
import com.jetbrains.php.lang.inspections.PhpInspection;
import com.jetbrains.php.lang.psi.PhpPsiElementFactory;
import com.jetbrains.php.lang.psi.elements.*;
import com.jetbrains.php.lang.psi.visitors.PhpElementVisitor;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.nextras.orm.intellij.utils.PhpClassUtils;
import org.nextras.orm.intellij.utils.PhpIndexUtils;

import java.util.Collection;

public class ReadOnlyPropertyInspection extends PhpInspection
{


	@NotNull
	@Override
	public String getShortName()
	{
		return "NextrasOrmReadOnlyProperty";
	}


	@NotNull
	@Override
	public PsiElementVisitor buildVisitor(@NotNull final ProblemsHolder problemsHolder, boolean b)
	{
		return new PhpElementVisitor()
		{
			@Override
			public void visitPhpAssignmentExpression(AssignmentExpression assignmentExpression)
			{
				if (!(assignmentExpression.getVariable() instanceof FieldReference)) {
					return;
				}
				if (!isReadOnlyProperty((FieldReference) assignmentExpression.getVariable())) {
					return;
				}
				problemsHolder.registerProblem(assignmentExpression, "Field is read only", new SetReadOnlyValueFix());
			}
		};
	}

	private boolean isReadOnlyProperty(FieldReference fieldReference)
	{
		PhpIndex phpIndex = PhpIndex.getInstance(fieldReference.getProject());
		Collection<PhpClass> classes = PhpIndexUtils.getByType(fieldReference.getClassReference().getType(), phpIndex);
		for (PhpClass cls : classes) {
			PhpClass entityInterface = PhpClassUtils.getInterface(phpIndex, "\\Nextras\\Orm\\Entity\\IEntity");
			if (!PhpClassUtils.isImplementationOfInterface(cls, entityInterface)) {
				continue;
			}
			Field field = cls.findFieldByName(fieldReference.getName(), false);
			if (field instanceof PhpDocProperty && field.getParent().getFirstChild().getText().equals("@property-read")) {
				return true;
			}
		}
		return false;
	}

	private class SetReadOnlyValueFix implements LocalQuickFix
	{

		@Nls
		@NotNull
		@Override
		public String getFamilyName()
		{
			return "Change to setReadOnlyValue";
		}

		@Override
		public void applyFix(@NotNull Project project, @NotNull ProblemDescriptor problemDescriptor)
		{
			AssignmentExpression assignment = (AssignmentExpression) problemDescriptor.getPsiElement();
			FieldReference fieldReference = (FieldReference) assignment.getVariable();

			MethodReference ref = PhpPsiElementFactory.createMethodReference(project, fieldReference.getClassReference().getText()
				+ "->setReadOnlyValue('" + fieldReference.getName() + "', "
				+ assignment.getValue().getText() + ")"
			);
			assignment.replace(ref);
		}
	}
}

File: src/org/nextras/orm/intellij/inspection/ReadOnlyPropertyInspection.kt
package org.nextras.orm.intellij.inspection

import com.intellij.codeInspection.LocalQuickFix
import com.intellij.codeInspection.ProblemDescriptor
import com.intellij.codeInspection.ProblemsHolder
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElementVisitor
import com.jetbrains.php.PhpIndex
import com.jetbrains.php.lang.documentation.phpdoc.psi.PhpDocProperty
import com.jetbrains.php.lang.inspections.PhpInspection
import com.jetbrains.php.lang.psi.PhpPsiElementFactory
import com.jetbrains.php.lang.psi.elements.AssignmentExpression
import com.jetbrains.php.lang.psi.elements.FieldReference
import com.jetbrains.php.lang.psi.visitors.PhpElementVisitor
import org.jetbrains.annotations.Nls
import org.nextras.orm.intellij.utils.OrmUtils
import org.nextras.orm.intellij.utils.PhpIndexUtils

class ReadOnlyPropertyInspection : PhpInspection() {
	override fun getShortName(): String {
		return "NextrasOrmReadOnlyProperty"
	}

	override fun buildVisitor(problemsHolder: ProblemsHolder, b: Boolean): PsiElementVisitor {
		return object : PhpElementVisitor() {
			override fun visitPhpAssignmentExpression(assignmentExpression: AssignmentExpression) {
				if (assignmentExpression.variable !is FieldReference) {
					return
				}
				if (!isReadOnlyProperty(assignmentExpression.variable as FieldReference)) {
					return
				}
				problemsHolder.registerProblem(assignmentExpression, "Field is read only", SetReadOnlyValueFix())
			}
		}
	}

	private fun isReadOnlyProperty(fieldReference: FieldReference): Boolean {
		val phpIndex = PhpIndex.getInstance(fieldReference.project)
		val classes = PhpIndexUtils.getByType(fieldReference.classReference!!.type, phpIndex)
		return classes
			.filter { OrmUtils.OrmClass.ENTITY.`is`(it, phpIndex) }
			.map { it.findFieldByName(fieldReference.name, false) }
			.any { it is PhpDocProperty && it.parent.firstChild.text == "@property-read" }
	}

	private inner class SetReadOnlyValueFix : LocalQuickFix {
		@Nls
		override fun getFamilyName(): String {
			return "Change to setReadOnlyValue"
		}

		override fun applyFix(project: Project, problemDescriptor: ProblemDescriptor) {
			val assignment = problemDescriptor.psiElement as AssignmentExpression
			val fieldReference = assignment.variable as FieldReference
			val ref = PhpPsiElementFactory.createMethodReference(
				project,
				"${fieldReference.classReference!!.text}->setReadOnlyValue('${fieldReference.name}', ${assignment.value!!.text})"
			)
			assignment.replace(ref)
		}
	}
}
